src/segment-loader.js

```diff
@@ -1047,13 +1049,31 @@ export default class SegmentLoader extends videojs.EventTarget {
     }
 
     let segmentInfo = this.pendingSegment_;
+    let segment = segmentInfo.segment;
+    let isWalkingForward = this.mediaIndex !== null;
 
     this.pendingSegment_ = null;
     this.recordThroughput_(segmentInfo);
-    this.mediaIndex = segmentInfo.mediaIndex;
-    this.fetchAtBuffer_ = true;
 
-    this.logger_('handleUpdateEnd_', this.mediaIndex);
+    this.state = 'READY';
+
+    // If we previously appended a segment that ends more than 3 targetDurations before
+    // the currentTime_ that means that our conservative guess was too conservative.
+    // In that case, reset the loader state so that we try to use any information gained
+    // from the previous request to create a new, more accurate, sync-point.
+    if (segment &&
+        segment.end &&
+        this.currentTime_() - segment.end > segmentInfo.playlist.targetDuration * 3) {
+      this.resetLoader();
+    } else {
+      this.mediaIndex = segmentInfo.mediaIndex;
+      this.fetchAtBuffer_ = true;
+    }
+
+    // Don't do a rendition switch unless the SegmentLoader is already walking forward
+    if (isWalkingForward) {
+      this.trigger('progress');
```

Comment:
I discussed this with Matt in person. Basically, there are trade-offs either way. With this in place we may not respond to bandwidth issues as quickly but we ensure that we are always in a stable playback mode before we do switch. This is less important for VOD but it's a different story for live. One problem is that `seekable` can become stalled because we are changing playlists without having sync. Another problem is that we can conservatively guess multiple times in a row and will end up fetching the same segment over and over at different renditions (worst-case being ping-ponging as Matt put it.) This change should always result in a sync point being established and some forward progress being made before a rendition is allowed to change. Timeout-caused rendition changes will still happen immediately so that the player can take emergency action.